File: scripts/_resize_one.py
"""Downscale a single image to MAX_EDGE (default 2048) using PIL.
Handles the JPEGs that sharp/libvips cannot decode on this machine."""
import sys
import os
from PIL import Image

file = sys.argv[1]
max_edge = int(sys.argv[2]) if len(sys.argv) > 2 else 2048

try:
    im = Image.open(file)
    im.load()
except Exception as e:
    print(f"fail {e}", file=sys.stderr)
    sys.exit(1)

w, h = im.size
edge = max(w, h)
if edge <= max_edge:
    print("keep")
    sys.exit(0)

before = os.path.getsize(file)

# Downscale with high-quality Lanczos
ratio = max_edge / float(edge)
new_size = (max(1, int(round(w * ratio))), max(1, int(round(h * ratio))))
im = im.convert("RGB") if im.mode not in ("RGB", "RGBA", "L", "P", "CMYK") else im
if im.mode == "P":
    im = im.convert("RGBA") if "transparency" in im.info else im.convert("RGB")
if im.mode == "CMYK":
    im = im.convert("RGB")
im = im.resize(new_size, Image.LANCZOS)

# Save to temp then swap (Windows-safe)
tmp = file + ".rs.tmp"
ext = os.path.splitext(file)[1].lower()
if ext == ".png":
    im.save(tmp, "PNG", optimize=True)
else:
    im.save(tmp, "JPEG", quality=85, optimize=True, progressive=False, subsampling=2)

after = os.path.getsize(tmp)
if after < before * 0.95:
    with open(tmp, "rb") as src, open(file, "wb") as dst:
        dst.write(src.read())
    print(f"ok {before // 1024}KB->{after // 1024}KB {w}x{h}->{new_size[0]}x{new_size[1]}")
else:
    print("nogain")
try:
    os.remove(tmp)
except OSError:
    pass
sys.exit(0)

File: scripts/_resize_one.js
#!/usr/bin/env node
/* Downscale a single image to MAX_EDGE (default 2048) if it's over-large.
   One file per process (fresh libvips) to dodge the Windows lock/state issue. */
'use strict';
const fs = require('fs');
const path = require('path');
const sharp = require('sharp');

const file = process.argv[2];
const MAX_EDGE = parseInt(process.argv[3] || '2048', 10);
if (!file) { console.error('usage: node _resize_one.js <file> [maxEdge]'); process.exit(1); }

(async () => {
  try {
    const meta = await sharp(file, { failOn: 'none' }).metadata();
    const edge = Math.max(meta.width || 0, meta.height || 0);
    if (edge <= MAX_EDGE) { console.log('keep'); process.exit(0); }
    const before = fs.statSync(file).size;
    const tmp = file + '.rs.tmp';
    const ext = path.extname(file).toLowerCase();
    let p = sharp(file, { failOn: 'none' }).rotate();
    if (meta.width >= meta.height) p = p.resize({ width: MAX_EDGE });
    else p = p.resize({ height: MAX_EDGE });
    if (ext === '.png') p = p.png({ compressionLevel: 9, effort: 10 });
    else p = p.jpeg({ quality: 82, mozjpeg: true, chromaSubsampling: '4:2:0' });
    await p.toFile(tmp);
    const after = fs.statSync(tmp).size;
    if (after < before * 0.95) {
      fs.writeFileSync(file, fs.readFileSync(tmp));
      console.log(`ok ${Math.round(before / 1024)}KB->${Math.round(after / 1024)}KB`);
    } else {
      console.log('nogain');
    }
    try { fs.unlinkSync(tmp); } catch (_) {}
    process.exit(0);
  } catch (e) {
    console.error(`fail ${e.message}`);
    process.exit(1);
  }
})();
